docs/spec/message-store-service-api.yml

```diff
@@ -0,0 +1,262 @@
+swagger: '2.0'
+info:
+  description: >-
+    A service that stores and manages blockchain events forwarded by one or more
+    Relayer services. The blockchain events are stored in the canonical data
+    format employed by Relayer components.
+  version: 0.1.0
+  title: Relayer Message Store Service
+tags:
+  - name: messages
+    description: >-
+      Represents blockchain events in the canonical data format used by
+      Relayers.
+  - name: events
+    description: Represents raw blockchain events emitted by a network.
+schemes:
+  - https
+paths:
+  '/messages/{id}':
+    put:
+      tags:
+        - messages
+      summary: Adds or updates a Relayer message representing a blockchain event.
+      description: >-
+        The endpoint is used to submit (and create) a message if it does not
+        already exist in the store. If it does, the existing message is updated
+        to include the additional proof data of the submitting Relayer. Only
```

Comment:
Adding proof information means that this component would need to understand the proof format. For example, for a Merkle Proof type of proof, there would be nothing to add, whereas for a threshold signed format the threshold signed data structure would need to be decoded, combined and re-encoded.

Reply:
Right, I think there are two approaches we can take here:
1. The Message Store Service understands the range of proof schemes that are employed by upstream components and is responsible for processing (e.g. aggregating threshold signatures) these proofs on behalf of clients consuming these messages.
2. The Message Store Service simply records proof objects submitted by different Relayers (if they are different), and consumers of these messages have to interpret and reason about the set of proofs associated with a given message. 

The current approach assumes the latter primarily because of the view that we should keep the Message Store service as simple and as minimally coupled to the mechanics of upstream components as we can. This will help this service cater to a wider range of use-cases and avoids delegating fundamental capabilities of the messaging layer (e.g. aggregation of threshold signatures) to a largely auxiliary component. Such capabilities could instead be achieved through the[inter-relay communication and coordination components](https://github.com/ConsenSys/gpact/blob/main/docs/relayer-design.md#architecture) in Relayers.

_Mechanics: Two messages regarding the same event, sent by different Relayers, should basically be identical except for (potentially) their proof payloads. The proof payload is an array of proof objects, where each object contains details and metadata about the specific proof. What this service would do is append any new (and different) proof data for a message to the set of proofs it already maintains for it._

Reply:
That makes sense and sounds like a good approach.